### ISSUE TITLE
M64P does not correctly implement N64's handling of denormalized floating-point numbers

### ISSUE THREAD

**Issue by bkimmett** [mupen64plus/mupen64plus-core#795]
The N64 hardware can't perform operations with [denormalized floating-point numbers](https://en.wikipedia.org/wiki/Denormal_number). Instead, it traps to software support, and many [possibly most/all] N64 games do not support this. 

However, the R4300i is _supposed to_ have a way to get around this: [If bit `01000000` is set in FCR31](https://hack64.net/docs/VR43XX.pdf#page=213), the N64 processor manual says that denormal numbers will be rounded away, to either 0 or the nearest normal floating-point number. 

I say 'supposed to' above because after performing some tests with an Everdrive, it turns out _the N64 does not actually respect whether this bit is set or not_; if it runs into a denormal number, it always throws an "Unimplemented Operation" exception - no matter what.

I'm bringing this up because M64P and its derivatives do not replicate this behavior. They either respect the `01000000` bit of FCR31 and round when the bit is set, or always round denormal numbers - I'm not sure which.

For emulation accuracy, it would be a good idea for M64P to at least offer the option to match the console's behavior, throwing "Unimplemented Operation" whenever a denormal floating-point number is found, regardless of the state of the `01000000` bit of FCR31.

**Comment by loganmc10** [comment]
mupen64plus in general doesn't worry about crashing games properly. Only a very small subnet of people find this to be useful behaviour, so there isn't a lot of developer interest in it.

We did sort out some issues with denormalized floats not long ago (https://github.com/mupen64plus/mupen64plus-core/commit/b7b56fea513e8734b982347d5b965c4a8e6f2074) which solved some issues with Rayman 2, but you're correct (as far as I'm aware) that throwing an exception isn't implemented properly.

Basically what I'm saying is that I don't think any commercial games wouldn't benefit from this, which is why it hasn't received much attention.

**Comment by bkimmett** [comment]
That's fair. I admit I found it while making a romhack, where it turned out an integer value was being passed to a function instead of a floating point value - and the int was interpreted as a denormalized float.
